src/traces/bar/plot.js

```diff
@@ -120,11 +122,13 @@ module.exports = function plot(gd, plotinfo, cdbar, barLayer) {
                 y1 = fixpx(y1, y0);
             }
 
-            Lib.ensureSingle(bar, 'path')
-                .style('vector-effect', 'non-scaling-stroke')
-                .attr('d',
-                    'M' + x0 + ',' + y0 + 'V' + y1 + 'H' + x1 + 'V' + y0 + 'Z')
-                .call(Drawing.setClipUrl, plotinfo.layerClipId);
+            if(!is_empty) {
+                Lib.ensureSingle(bar, 'path')
+                    .style('vector-effect', 'non-scaling-stroke')
+                    .attr('d',
+                        'M' + x0 + ',' + y0 + 'V' + y1 + 'H' + x1 + 'V' + y0 + 'Z')
+                    .call(Drawing.setClipUrl, plotinfo.layerClipId);
+            }
```

Comment:
Then we also need to remove the bar if it later disappears. Off your branch I can do:
```js
Plotly.newPlot(gd,
    [{type:'bar',y:[0,1,0,1],text:'hi',textposition:'outside',cliponaxis:false}],
    {width: 400, height: 400})
```
Which looks great:
<img width="290" alt="screen shot 2018-10-29 at 11 14 56 am" src="https://user-images.githubusercontent.com/2678795/47659738-f7317f80-db6b-11e8-86b1-dcb8c2fb7624.png">
But then if I do:
```js
Plotly.restyle(gd,'y',[[1,0,1,0]])
```
The newly-removed bars are still there (with their text moved to the right place):
<img width="280" alt="screen shot 2018-10-29 at 11 16 12 am" src="https://user-images.githubusercontent.com/2678795/47659814-20521000-db6c-11e8-9d99-72f695beb035.png">